### ISSUE TITLE
Error in load/save models with FeatureLayer when refitting or reloading

### ISSUE THREAD

**Issue by kglapiak** [tensorflow/tensorflow#39221]
Hello,

My model has FeatureLayer and i want do flow like this:

1. Create, fit model  on basic data and save.
2. Load model and fit on new data and save updated model.
3. Do 2 step periodically e.g. 1 time in week.

When I do model  first time, it's no problem, also I can use it in tensorflow-serving and it works.
Then I can load this model, fit on new data and save updated model. But when I try to load updated model i have an error:

` raise ValueError('Passing a dictionary input to a Sequential Model '
ValueError: Passing a dictionary input to a Sequential Model which doesn't have FeatureLayer as the first layer is an error.`

In tf v 2.0.0 problem is when load model:
```
model = tf_load.load_internal(path, loader_cls=KerasObjectLoader)
File "/usr/lib64/python3.6/site-packages/tensorflow_core/python/saved_model/load.py", line 541, in load_internal
export_dir)
File "/usr/lib64/python3.6/site-packages/tensorflow_core/python/keras/saving/saved_model/load.py", line 103, in __init__
self._finalize()
File "/usr/lib64/python3.6/site-packages/tensorflow_core/python/keras/saving/saved_model/load.py", line 132, in _finalize
node._set_inputs(inputs)
File "/usr/lib64/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py", line 2697, in _set_inputs
inputs = self._set_input_attrs(inputs)
File "/usr/lib64/python3.6/site-packages/tensorflow_core/python/training/tracking/base.py", line 457, in _method_wrapper
result = method(self, *args, **kwargs)
File "/usr/lib64/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py", line 2731, in _set_input_attrs
raise ValueError('Passing a dictionary input to a Sequential Model '
ValueError: Passing a dictionary input to a Sequential Model which doesn't have FeatureLayer as the first layer is an error.
```
I tested it and it's not problem with fitting model but with save or load, becouse when I only load first model and save as new model (without any interference in model) the error occurs when i try load updated model.

In version 2.1.0 its not error when load updated model, but when I try to fit. Final error is the same in both cases
```
 File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\training.py", line 819, in fit
    use_multiprocessing=use_multiprocessing)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\training_v2.py", line 235, in fit
    use_multiprocessing=use_multiprocessing)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\training_v2.py", line 593, in _process_training_inputs
    use_multiprocessing=use_multiprocessing)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\training_v2.py", line 706, in _process_inputs
    use_multiprocessing=use_multiprocessing)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\data_adapter.py", line 702, in __init__
    x = standardize_function(x)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\training_v2.py", line 660, in standardize_function
    standardize(dataset, extract_tensors_from_dataset=False)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\training.py", line 2346, in _standardize_user_data
    all_inputs, y_input, dict_inputs = self._build_model_with_inputs(x, y)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\training.py", line 2572, in _build_model_with_inputs
    self._set_inputs(cast_inputs)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\training.py", line 2647, in _set_inputs
    inputs = self._set_input_attrs(inputs)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\training\tracking\base.py", line 457, in _method_wrapper
    result = method(self, *args, **kwargs)
  File "C:\Users\kglapiak\AppData\Local\Continuum\anaconda3\envs\tf_test2\lib\site-packages\tensorflow_core\python\keras\engine\training.py", line 2681, in _set_input_attrs
    raise ValueError('Passing a dictionary input to a Sequential Model '
ValueError: Passing a dictionary input to a Sequential Model which doesn't have FeatureLayer as the first layer is an error.
```

I save model by mehod:
`model.save('original_model',save_format='tf')`

Example of model:
```
feature_columns = []

categorical = tf.feature_column.categorical_column_with_vocabulary_list(
        key='f1', vocabulary_list=['x1','x2','x3'],
        num_oov_buckets=0)
one_hot = feature_column.indicator_column(categorical)
feature_columns.append(one_hot)

categorical = tf.feature_column.categorical_column_with_vocabulary_list(
        key='f2', vocabulary_list=['x1','x2','x3','x4','x5','x6'],
        num_oov_buckets=5)
embedding = feature_column.embedding_column(categorical, dimension=10)
feature_columns.append(embedding)

feature_layer = tf.keras.layers.DenseFeatures(feature_columns)

model = tf.keras.Sequential([
  feature_layer,
  layers.Dense(20, activation='relu'),
  layers.Dense(1,activation='softsign')
])
    
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])


def df_to_dataset(dataframe, shuffle=True, batch_size=32):
    dataframe = dataframe.copy()
    labels = dataframe.pop('target')
    ds = tf.data.Dataset.from_tensor_slices((dict(dataframe), labels))
    if shuffle:
        ds = ds.shuffle(buffer_size=len(dataframe))
    ds = ds.batch(batch_size)
    return ds

train = df_to_dataset(train)
val = df_to_dataset(val)

model.fit(train,
          validation_data=val,
          epochs=2)
```

I tested it on conda and pip distributions, on tf versions 2.0.0 and 2.1.0 and on Windows and Centos and several versions of Python and all time is the same problem.

- OS Platform and Distribution (Windows, Centos):
- TensorFlow version (2.0.0,2.1.0 also pip and conda distridutions ):
- Python version: several versions from  3.6.5 to 3.7.1

**Comment by Saduf2019** [comment]
@kglapiak 
Code shared is incomplete we will not be able to replicate the issue, please share simple stand alone code or if possible a colab gist for us to analyse the issue faced

**Comment by kglapiak** [comment]
All code:

```
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

import tensorflow as tf
from tensorflow.keras.layers import Activation, Dropout, Flatten, Dense
from tensorflow.keras.models import Sequential
from tensorflow.keras.models import load_model,Model
from tensorflow import feature_column
from tensorflow.keras import layers


def df_to_dataset(dataframe, shuffle=True, batch_size=32):
  dataframe = dataframe.copy()
  labels = dataframe.pop('target')
  ds = tf.data.Dataset.from_tensor_slices((dict(dataframe), labels))
  if shuffle:
    ds = ds.shuffle(buffer_size=len(dataframe))
  ds = ds.batch(batch_size)
  return ds

###--------- sample dataset ---------
n = 10000
data = pd.DataFrame({'f1':np.random.choice(['x1','x2','x3'], n, replace=True),
                     'f2':np.random.choice(['x1','x2','x3','x4','x5','x6'], n, replace=True),
                     'target':np.random.choice([0,1], n, replace=True)})

train,val = train_test_split(data,test_size=0.2)
train,new_train = train_test_split(train,test_size=0.2)


###--------- define features columns ---------
feature_columns = []

categorical = tf.feature_column.categorical_column_with_vocabulary_list(
        key='f1', vocabulary_list=['x1','x2','x3'],
        num_oov_buckets=0)
one_hot = feature_column.indicator_column(categorical)
feature_columns.append(one_hot)

categorical = tf.feature_column.categorical_column_with_vocabulary_list(
        key='f2', vocabulary_list=['x1','x2','x3','x4','x5','x6'],
        num_oov_buckets=5)
embedding = feature_column.embedding_column(categorical, dimension=10)
feature_columns.append(embedding)

feature_layer = tf.keras.layers.DenseFeatures(feature_columns)

###--------- define and compile model ---------
model = tf.keras.Sequential([
  feature_layer,
  layers.Dense(20, activation='relu'),
  layers.Dense(1,activation='softsign')
])
    
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])


train = df_to_dataset(train)
val = df_to_dataset(val)
new_train = df_to_dataset(new_train)

#----------- train model first time - all its ok --------
model.fit(train,
          validation_data=val,
          epochs=2)

model.save('basic_model',save_format='tf')

###------------- first update, no errors -------------------------
model_update1 = tf.keras.models.load_model('basic_model')
model_update1.fit(new_train,
          validation_data=val,
          epochs=2)

model_update1.save('update1',save_format='tf')

###------------- 2nd update is problem and generate errors -------------------------
model_update2 = tf.keras.models.load_model('update1')
model_update2.fit(new_train,
          validation_data=val,
          epochs=2)

```

**Comment by Saduf2019** [comment]
@kglapiak 
I ran the code shared above on tf 2.1 but do not face any error as reported above, please find the [gist here](https://colab.sandbox.google.com/gist/Saduf2019/2b40a2625e180590b2ff6ec028dac7ac/2.ipynb)

**Comment by kglapiak** [comment]
For me this colab returns error:

```
ValueError                                Traceback (most recent call last)
<ipython-input-4-b1ad4c6f08c2> in <module>()
     82 model_update2.fit(new_train,
     83           validation_data=val,
---> 84           epochs=2)

10 frames
/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training.py in _set_input_attrs(self, inputs)
   2679         # We assert that the first layer is a FeatureLayer.
   2680         if not training_utils.is_feature_layer(self.layers[0]):
-> 2681           raise ValueError('Passing a dictionary input to a Sequential Model '
   2682                            'which doesn\'t have FeatureLayer as the first layer'
   2683                            ' is an error.')

ValueError: Passing a dictionary input to a Sequential Model which doesn't have FeatureLayer as the first layer is an error.
```

**Comment by pcejrowski** [comment]
Correct @kglapiak.
I've just tried taking a look at this issue, and yes -- it's breaking up exactly where you've pointed out.
@Saduf2019  - have you tested out the linked gist?

**Comment by Saduf2019** [comment]
@kglapiak 
I ran the code shared by you and do not face any error, in case you are still facing an error can you please share a colab gist so we could analyse the error faced. faced

**Comment by kglapiak** [comment]
Here colab [gist](https://colab.research.google.com/drive/17h5_Tceuh8Pt7mjRdaJ6xOPZY6XovaIr?usp=sharing) exacly same like yours. 
And the same error occurs.
I show this for some people and they all had the same.
If you are sure, its works for you, only explanation is it's work in US and dot't worki in EU.
Is it even possible?

**Comment by Saduf2019** [comment]
@kglapiak
Could you please give me permission to view the gist.

**Comment by kglapiak** [comment]
@Saduf2019 yea - now should be ok

**Comment by Saduf2019** [comment]
I am able to replicate the reported issue on tf 2.1,please find the [gist here](https://colab.sandbox.google.com/gist/Saduf2019/3099abb57ef7134f37949a6b36d99f74/untitled177.ipynb). Thanks!

**Comment by kglapiak** [comment]
Hi, @ymodak - something happends with that issue?

**Comment by ymodak** [comment]
This is fixed with TensorFlow 2.2
Thanks!

**Comment by kglapiak** [comment]
Yea, its works. Thanks !
